### ISSUE TITLE
Feature request - AI-Generated commit message feature similar to github copilot in VS Code

### ISSUE THREAD

**Issue by vgibilmanno** [ForkIssues/TrackerWin#2100]
In Visual Studio Code, there's a feature with GitHub Copilot that automatically generates commit messages. This feature is quite helpful as it saves time and provides relevant commit messages.

I think it would be beneficial to have a similar feature in fork. An AI that suggests commit messages based on the changes in a commit would be a significant improvement. It’s a small enhancement that can have a big impact on daily workflow, especially for those handling numerous commits.

See VS Code:
![image](https://github.com/fork-dev/TrackerWin/assets/19838802/4b276d60-130c-4012-be00-a836411b42af)

**Comment by DanPristupov** [comment]
Does it require a paid subscription?

**Comment by vgibilmanno** [comment]
Yes you need a paid subscription. This is their subscription model:

> GitHub Copilot can be managed through personal accounts with GitHub Copilot Individual or through organization accounts with GitHub Copilot Business.

> GitHub Copilot is free to use for verified students, teachers, and maintainers of popular open source projects. If you are not a student, teacher, or maintainer of a popular open source project, you can try GitHub Copilot for free with a one-time 30-day trial. After the free trial, you will need a paid subscription for continued use. For more information, see "[About billing for GitHub Copilot](https://docs.github.com/en/billing/managing-billing-for-github-copilot/about-billing-for-github-copilot)."

The question is, if it has to be github copilot. It's up to you/your team what kind of AI you would want to integrate. Personally I would prefer github copilot though

Edit: I noticed that Github Copilot doesn't offer an API

**Comment by skttl** [comment]
The GitLens extension for VSCode also does this, and lets you provide an API key for OpenAI and select which model to use

https://github.com/gitkraken/vscode-gitlens/blob/ebcbf05d5a83479a71dc12d78b23d49e15713f78/src/commands/generateCommitMessage.ts#L17

**Comment by DefinitelyADev** [comment]
This would be awesome!

**Comment by yotamcuralife** [comment]
For even more convenience, How can I make it automatically generate a commit message when i push on the commit button?

**Comment by javadibrahimli** [comment]
I think its performance is not too good, cause, it does a lot of repetitions

**Comment by edelciomolina** [comment]
I recently submitted [this idea](https://github.com/fork-dev/TrackerWin/issues/2322) for use with GPT;
But thinking a little deeper, I created a Custom Command with a target for File, see

![image](https://github.com/user-attachments/assets/6a28905c-8f89-44d7-96a7-a5992ff24ae3)

![image](https://github.com/user-attachments/assets/b781cd5e-4d57-49f0-8c56-836f06ec8ce8)

And I used the following script, which is from using [the chatgpt NPM package](https://www.npmjs.com/package/chatgpt):

```sh
npx chatgpt@latest --model gpt-3.5-turbo "Based on the following text, create 3 standardized message options for GIT Command including at beginning of messate a gitmoji icon (show only the text, not commands): $1{text}" --apiKey sk-your-gpt-apikey
```
> You can use gpt-4 as a model, but I preferred gpt-3.5 because it seemed faster to me.

So, in the text box you can write whatever you prefer and the output will be the suggestions. See the example below:

```
I made a correction to the G Suite import system and improved the entire data validation command.
```

Response:
```
🔧 Made a correction in the gsuite import system and improved the data validation process.

🚀 Fixed the gsuite import system and enhanced data validation commands.

🛠️ Improved data validation commands after fixing the gsuite import system.
``` 

## Future Idea for Custom Commands
- A new target called "Commit Button" _(That can be a button on side of Commit dropdown button at below the commit description)_
- Create keywords $subject and $description for get the texts input
- A checkbox "Replace Subject with the output response"
- A checkbox "Replace Description with the output response"

**Comment by igloo15** [comment]
It would be nice if you could just alter the git commit command. If you could customize that it would solve most problems. For instance [OpenCommit](https://github.com/di-sukharev/opencommit) is a good tool for AI generated commit messages. It works by using its own cli to proxy git. So `opencommit` or `oco` replaces `git commit -m "message here"`. The cli does the commit and generates the message while it does it. So if you could customize/change the git command for committing you could easily integrate this tool and others.

**Comment by Skoddie** [comment]
[SourceGit](https://github.com/sourcegit-scm/sourcegit) also has a nice implementation that can also be used with LLMs running locally via Ollama:

![Image](https://github.com/user-attachments/assets/4bffda98-e24f-487d-9afa-59998d6a3322)

![Image](https://github.com/user-attachments/assets/5c94a78d-1159-4409-9c0a-ad7e6c13a094)

**Comment by Northwind877** [comment]
Such a feature would be brilliant in Fork! I like to be lazy with the git messages :)

**Comment by exytab** [comment]
Copilot Language Server SDK is now available

https://github.blog/changelog/2025-02-10-copilot-language-server-sdk-is-now-available/

**Comment by mrc-eu** [comment]
Please don't force AI into my workflow. If you add AI, then please as an option, so that I can turn it off (immediately).

**Comment by graphnode** [comment]
To make it future proof it would be better to allow the git message to be generated from a command. Like having a button near the message box which calls an action similar to what we already have for branches, commits, repositories, etc the output of that command would then be inserted into the message box for the user to check before commit.

This way this would work for AI and other tools without getting vendor locked.

**Comment by PadreSVK** [comment]
+1 for ability to generate message from cmd instead actual OpenAI only solution :) 

<img width="586" height="143" alt="Image" src="https://github.com/user-attachments/assets/2c491772-e3a2-4572-b53b-d128081e387f" />

**Comment by DanPristupov** [comment]
In Fork 2.14 we added support for Claude and will add Codex in the future. OpenAI is going to be deprecated.

**Comment by PadreSVK** [comment]
@DanPristupov what about gh copilot? :)

**Comment by graphnode** [comment]
> In Fork 2.14 we added support for Claude and will add Codex in the future. OpenAI is going to be deprecated.

> [@DanPristupov](https://github.com/DanPristupov) what about gh copilot? :)

This is why I mentioned the command idea. It's going to be pain to keep track every new shiny AI that shows up and any changes that happen to their API.

**Comment by mAcf00bar** [comment]
> In Fork 2.14 we added support for Claude and will add Codex in the future. OpenAI is going to be deprecated.

I'm on 2.14.1.0 and I can only see the OpenAI login button.
As a suggestion (and to avoid to always have the latest implemented), how about doing it like Linqpad does:

<img width="752" height="569" alt="Image" src="https://github.com/user-attachments/assets/e88ecdd0-a3d6-495d-85ec-22da5179659d" />

Here you would be able to setup your own AI Provider:

<img width="302" height="520" alt="Image" src="https://github.com/user-attachments/assets/244b98bc-e269-4f08-80d8-e971070ec624" />

**Comment by VIEWVIEWVIEW** [comment]
> In Fork 2.14 we added support for Claude and will add Codex in the future. OpenAI is going to be deprecated.

Any update on adding Codex commit messages on Windows? @DanPristupov

**Comment by vlabo** [comment]
At my company we have a policy which LLMs we can use. And there is also a company hosted one.
My current solution is to have a custom command that generates commit message and I manually copy past it.
This is not a big issue, but it will be nice if it can forward the output of the custom command to the commit message ui element.

**Comment by edelciomolina** [comment]
It would be great to have a language option in the Commit tab preferences.
<kbd>🇧🇷</kbd> <kbd>🇪🇸</kbd> <kbd>🇺🇸</kbd> <kbd>🇫🇷</kbd>
